### ISSUE TITLE
Color.toString() for Safari Canvas?

### ISSUE THREAD

**Issue by mattdesl** [LeaVerou/color.js#304]
I'm using Color.js (thanks!) and it seems like `color.toString()` doesn't always work in Safari for Canvas2D, which expects colors to be in `"rgb(x, y, z)"` — although `"color(display-p3 1 0 2)"` does work. But colors like `"lch(100 50 40)"` produce black pixels.

I'm not sure the root cause of this but it appears like it may be a browser bug? Chrome renders fine, and Safari Technology Preview renders fine too (Release 168, Safari 16.4, WebKit 18616.1.10.2).

Problematic Device/Browser:
MacBook Air M2, 2022, Ventura 13.0 (22A380)
Safari Version 16.1 (18614.2.9.1.12)

For now I have the following code that I am using before sending the color to `context.fillStyle`

```js
const supportsP3 =
  window.CSS && CSS.supports("color", "color(display-p3 0 1 0)");
const outputSpaceId = supportsP3 ? "p3" : "srgb";

context.fillStyle = colorToStyle(new Color("lch(100 50 40)"));

export function colorToStyle(color) {
  // re-parse string in case it's out of gamut
  if (typeof color === "string") {
    return colorToStyle(new Color(color));
  }

  // edge case
  if (!color || !color.spaceId) return String(color);

  // operate on a copy
  color = color.clone();

  // convert to output space
  if (!color.spaceId !== outputSpaceId) color = color.to(outputSpaceId);

  // clamp to space's color gamut
  color.toGamut({ space: outputSpaceId });

  return color.toString();
}
```

In general, it would be nice to have a sort of "toStyle()" or "toHex()" that does something like this, automatically converting the color to a simpler RGB space and then rendering it as a full hex (with alpha if necessary) or old-school RGB string. This would be useful for legacy browsers and environments that don't support the full CSS color spec.

EDIT: Slightly improved my function to support a wider gamut when input color is something like oklch. This issue seems related somewhat to #31

**Comment by LeaVerou** [comment]
`display()` does exactly this, though it uses `CSS.supports()` under the hood, we had not considered that Canvas may not support all the CSS colors supported by the browser. One solution would be to pass in a parameter to indicate the intended usage of the color and use a dummy canvas if `target: "canvas"` to test if it's supported, however that would be quite slow.

**Comment by svgeesus** [comment]
Canvas 2D context can be created with [either an sRGB or a display-p3 canvas](https://html.spec.whatwg.org/dev/canvas.html#2dcontext). However it is color managed, and [any CSS color](https://html.spec.whatwg.org/dev/canvas.html#fill-and-stroke-styles) can be specified for drawing (fill, stroke); the color is converted to the canvas color space on input. So this looks like a browser bug on older versions. The [16.4 release notes](https://developer.apple.com/documentation/safari-release-notes/safari-16_4-release-notes) don't mention anything specific to this bug, and I checked back to 16.0 without finding mention of it either. But I remember that very old Safari versions would only accept either sRGB or display-p3 in CSS; the rest was added later as CSS Color 4 matured and was fully implemented.

As @LeaVerou says, `display()` does what you want here.
